### PR TITLE
[New Integration] Add ReConfirm integration

### PR DESCRIPTION
**Pull request by m11tch**
Suggested label: Enhancement

## Proposed commit message

[New Integration] Add ReConfirm integration

Adds the ReConfirm integration to collect attack-surface and vulnerability
scan results via webhook and normalize them to ECS.

WHAT:
- Adds a new `vulnerability` data stream backed by an `http_endpoint` input;
  ReConfirm POSTs one JSON document per completed scan.
- Reshapes ReConfirm's dynamically keyed sections, including assets by IP,
  subdomains by hostname, vulnerability findings, associated domains, and
  credential leaks, into bounded nested arrays instead of unbounded dynamic
  mappings.
- Maps vulnerability findings onto ECS `vulnerability.*` fields, including
  severity, CVE ID, and CVSS score, alongside ReConfirm-specific fields under
  `reconfirm.*`.
- Adds `redact_leaked_passwords`, enabled by default, to replace leaked
  plaintext credential values with `REDACTED` in both
  `reconfirm.cred_leaks.password` and `event.original`.

WHY:
- Lets security teams track external attack-surface findings, exposed
  services, lookalike domains, leaked credentials, and email security posture
  in Elasticsearch alongside their other telemetry.

## Checklist

- [x] I have reviewed [tips for building integrations](https://www.elastic.co/docs/extend/integrations/tips-for-building) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [x] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices)

N/A: no dashboards are included in this version.

## Author's Checklist

- [x] Update `changelog.yml`'s PR link from the `pull/1` placeholder to this PR's actual number once opened.

## How to test this PR locally

```bash
cd packages/reconfirm
elastic-package build
elastic-package format --fail-fast
elastic-package lint
elastic-package check
elastic-package test pipeline
elastic-package test system --generate
elastic-package test system
```

All commands passed locally.

## Related issues

-

## Screenshots

N/A: no dashboards or visual UI assets are included in this version.

### CODE REVIEW

**Comment by qcorporation** [comment]
@vera-review-bot review

File: packages/reconfirm/manifest.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  - name: reconfirm
+    title: ReConfirm scan results
+    description: Collect attack-surface vulnerability scan results from ReConfirm.
+    inputs:
+      - type: http_endpoint
+        title: Collect ReConfirm scan results via webhook
+        description: Receive scan result webhooks pushed by ReConfirm on scan completion.
+owner:
+  github: elastic/security-service-integrations
+  type: community
```
[the remark this answers is not in the document]
Comment (m11tch) [reply packages/reconfirm/manifest.yml]: @qcorporation I wasn't entirely sure on what to pick here, can you confirm this is the right team/owner?

File: packages/reconfirm/data_stream/vulnerability/agent/stream/http_endpoint.yml.hbs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+publisher_pipeline.disable_host: true
+{{/contains}}
+
+tracer:
+  enabled: {{enable_request_tracer}}
+  filename: "../../logs/http_endpoint/http-request-trace-*.ndjson"
+  maxbackups: 5
+
+processors:
+  - add_fields:
```
[the remark this answers is not in the document]
Comment (m11tch) [reply packages/reconfirm/data_stream/vulnerability/agent/stream/http_endpoint.yml.hbs]: Fixed in 817962289b — de-indented the built-in `add_fields` processor to column 0 to match where `{{processors}}` renders, following the same pattern used in `packages/bitdefender/data_stream/push_notifications`.

File: packages/reconfirm/manifest.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (m11tch) [reply packages/reconfirm/manifest.yml]: going with `type: elastic` in fa32af5c51 per the bot's reasoning — owner.github is already elastic/security-service-integrations and every other package under that team uses `type: elastic`. Flagging in case there's a reason it should stay `community` that I'm missing.

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+              failed with message '{{{ _ingest.on_failure_message }}}'
+  # Fall back to ingest time when info.end is absent or unparseable.
+  - set:
+      field: '@timestamp'
+      tag: set_timestamp_fallback
+      value: '{{{_ingest.timestamp}}}'
+      if: ctx['@timestamp'] == null
+  - append:
+      field: error.message
+      tag: append_timestamp_fallback_note
```
[the remark this answers is not in the document]
Comment (m11tch) [reply packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml:107]: Fixed in 7ccbcc94bb — moved the note to `tags` (`reconfirm_timestamp_from_ingest_time`) instead of `error.message`, so these documents keep `event.kind: event`. Also added a dedicated test fixture (`scan-error-sample`), since none of the existing ones actually exercised a null `info.end`.

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+            item.ports = reshapePorts(a.ports);
+            item.user_tags = a.user_tags;
+            item.vulns = reshapeVulns(a.vulns, false);
+            assetsArr.add(item);
+          }
+          reconfirm.assets = assetsArr;
+          if (relatedIps.length > 0) {
+            ctx.related = ctx.related == null ? new HashMap() : ctx.related;
+            def ri = ctx.related.ip == null ? new ArrayList() : ctx.related.ip;
+            ri.addAll(relatedIps);
```
[the remark this answers is not in the document]
Comment (m11tch) [reply packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml]: Fixed in 9e67bce980 — added a `mergeUnique` helper and used it at all three `related.ip` merge sites. Also updated the happy-path fixture to include a genuinely overlapping IP (`198.51.100.10`, both an asset and a subdomain address) to cover it.

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+                continue;
+              }
+              def item = new HashMap();
+              item.email = entry.getKey();
+              item.password = redactLeakedPasswords && leak.password != null ? 'REDACTED' : leak.password;
+              item.sources = leak.sources;
+              item.last_breach = leak.last_breach;
+              clArr.add(item);
+            }
+            relatedLeakedEmails.add(entry.getKey());
```
[the remark this answers is not in the document]
Comment (m11tch) [reply packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml]: Fixed in afcfd36dd6 — moved `relatedLeakedEmails.add` inside the inner loop so it only fires when a leak entry actually survives the `Map` guard. The existing malformed-shapes fixture already covers this case.

File: packages/reconfirm/data_stream/vulnerability/_dev/test/pipeline/test-common-config.yml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+fields:
+  tags:
+    - preserve_original_event
+  _tmp:
+    redact_leaked_passwords: true
```
[the remark this answers is not in the document]
Comment (m11tch) [reply packages/reconfirm/data_stream/vulnerability/_dev/test/pipeline/test-common-config.yml:5]: Fixed in c6f06f5f1c — added `test-reconfirm-vulnerability-cred-leaks-unredacted-sample.json` with a per-test config disabling redaction, asserting the plaintext password survives.

File: packages/reconfirm/manifest.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (qcorporation) [reply packages/reconfirm/manifest.yml]: I believe that's correct, @narph @jamiehynds can you help me out with team ownership here?

**Comment by qcorporation** [comment]
@m11tch I've also added the @elastic/security-service-integrations team as a reviewer to get more eyes on this contribution (thank you btw)

**Comment by qcorporation** [comment]
@vera-review-bot review

File: packages/reconfirm/_dev/build/docs/README.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
+{{- generatedHeader }}
+# ReConfirm Integration for Elastic
+
+## Overview
+
+[ReConfirm](https://reconfirm.com/) is an attack-surface and vulnerability scanning service. The ReConfirm integration for Elastic collects scan result webhooks (assets, open ports, lookalike/typosquat domains, subdomain exposure, and email security posture), normalizes the scan envelope to the Elastic Common Schema (ECS), and ships it to Elasticsearch so security teams can track their external attack surface over time.
+
+### Compatibility
+
+This integration is compatible with ReConfirm's webhook notifications and Elastic Stack 8.19 and later.
```
Comment (efd6) [inline packages/reconfirm/_dev/build/docs/README.md]: Do we have a version for ReConfirm?

We don't need to note the stack version compatibility since this is included in the manifest and including it here brings the risk of staling.

File: packages/reconfirm/data_stream/vulnerability/agent/stream/http_endpoint.yml.hbs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+tracer:
+  enabled: {{enable_request_tracer}}
+  filename: "../../logs/http_endpoint/http-request-trace-*.ndjson"
+  maxbackups: 5
+
+processors:
+- add_fields:
+    target: ''
+    fields:
+      _tmp:
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/agent/stream/http_endpoint.yml.hbs]: This seems like a conf more than a temp.

File: packages/reconfirm/data_stream/vulnerability/_dev/test/pipeline/test-reconfirm-vulnerability-cred-leaks-unredacted-sample.json-expected.json [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                    "info"
+                ]
+            },
+            "observer": {
+                "product": "Attack Surface Scan",
+                "vendor": "ReConfirm"
+            },
+            "reconfirm": {
+                "assets": [],
+                "associated_domains": [],
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/_dev/test/pipeline/test-reconfirm-vulnerability-cred-leaks-unredacted-sample.json-expected.json]: Please add a script to the ingest pipeline to remove empty values. For example
```
  - script:
      description: This script processor iterates over the whole document to remove fields with null values.
      tag: remove_nulls_0370f4ef
      source: |-
        void handleMap(Map map) {
          map.values().removeIf(v -> {
            if (v instanceof Map) {
                handleMap(v);
            } else if (v instanceof List) {
                handleList(v);
            }
            return v == null || v == '' || (v instanceof Map && v.size() == 0) || (v instanceof List && v.size() == 0)
          });
        }
        void handleList(List list) {
          list.removeIf(v -> {
            if (v instanceof Map) {
                handleMap(v);
            } else if (v instanceof List) {
                handleList(v);
            }
            return v == null || v == '' || (v instanceof Map && v.size() == 0) || (v instanceof List && v.size() == 0)
          });
        }
        handleMap(ctx);
```

File: packages/reconfirm/data_stream/vulnerability/_dev/test/pipeline/test-reconfirm-vulnerability-malformed-port-key-sample.json-expected.json [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                "ip": [
+                    "198.51.100.30",
+                    "203.0.113.30"
+                ],
+                "user": [
+                    "jane.doe@example.com"
+                ]
+            },
+            "source": {
+                "ip": "203.0.113.30"
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/_dev/test/pipeline/test-reconfirm-vulnerability-malformed-port-key-sample.json-expected.json:86]: Suggest doing geoip look ups for `source.ip`.

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml [hunk 11]
```diff
@@ -0,0 +1,620 @@
...
+      target_field: event.original
+      ignore_missing: true
+      description: Renames the original `message` field to `event.original` to store a copy of the original message. The `event.original` field is not touched if the document already has one; it may happen when Logstash sends the document.
+      if: ctx.event?.original == null
+  - remove:
+      field: message
+      tag: remove_message
+      ignore_missing: true
+      description: The `message` field is no longer required if the document has an `event.original` field.
+      if: ctx.event?.original != null
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml:19]: Why are we doing this? The input is http_endpoint, so message is never set.

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml [hunk 12]
```diff
@@ -0,0 +1,620 @@
...
+  # email_security has a fixed shape (dkim/dmarc/mx/spf), safe to pass through
+  # as-is without reshaping.
+  ##
+  - rename:
+      field: json.results.email_security
+      tag: rename_email_security
+      target_field: reconfirm.email_security
+      ignore_missing: true
+  # Reshape dynamic-keyed result objects into arrays to keep mappings bounded.
+  - script:
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml:178]: Can we break this up into specific functionality rather than putting everything in one script.

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml [hunk 13]
```diff
@@ -0,0 +1,620 @@
...
+          } else if (o instanceof Map) {
+            ((Map) o).values().removeIf(v -> drop(v));
+            return (((Map) o).size() == 0);
+          } else if (o instanceof List) {
+            ((List) o).removeIf(v -> drop(v));
+            return false;
+          }
+          return false;
+        }
+        drop(ctx);
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml:834]: From the test results, this does not work.

File: packages/reconfirm/data_stream/vulnerability/manifest.yml [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+      - name: url
+        type: text
+        title: "Webhook path"
+        description: URL path where the webhook accepts requests. It must include a leading slash.
+        show_user: true
+        default: /reconfirm
+      - name: data_stream.dataset
+        type: text
+        title: "Dataset name"
+        description: |-
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/manifest.yml]:
```suggestion
        description: >-
```

File: packages/reconfirm/data_stream/vulnerability/manifest.yml [hunk 15]
```diff
@@ -0,0 +1,108 @@
+title: "ReConfirm Vulnerability Scan Results"
+type: logs
+dataset: reconfirm.vulnerability
+streams:
+  - input: http_endpoint
+    title: "ReConfirm scan results via webhook"
+    description: |-
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/manifest.yml]:
```suggestion
    description: >-
```

File: packages/reconfirm/data_stream/vulnerability/manifest.yml [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        type: password
+        title: "Authorization Token"
+        description: Exact value expected in the incoming `Authorization` header, for example `Bearer <token>`.
+        required: true
+        secret: true
+        show_user: true
+      - name: preserve_original_event
+        type: bool
+        title: "Preserve Original Event"
+        description: |-
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/manifest.yml]:
```suggestion
        description: >-
```

File: packages/reconfirm/data_stream/vulnerability/manifest.yml [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+          This option copies the raw unmodified body of the incoming request to the
+          event.original field as a string before sending the event to Elasticsearch.
+        show_user: true
+        default: false
+      - name: redact_leaked_passwords
+        type: bool
+        title: "Redact Leaked Passwords"
+        description: |-
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/manifest.yml]:
```suggestion
        description: >-
```

File: packages/reconfirm/data_stream/vulnerability/manifest.yml [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+        title: "TLS"
+        description: TLS options for the listener. Leave disabled when TLS is terminated before Elastic Agent. See [documentation](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-ssl.html) for all options.
+        default: |
+          enabled: false
+          certificate: "/etc/pki/client/cert.pem"
+          key: "/etc/pki/client/cert.key"
+      - name: processors
+        type: yaml
+        title: "Processors"
+        description: |-
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/manifest.yml]:
```suggestion
        description: >-
```

File: packages/reconfirm/data_stream/vulnerability/manifest.yml [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+          Processors are used to reduce the number of fields in the exported event or
+          to enhance the event with metadata. This executes in the agent before the
+          logs are parsed. See
+          [Processors](https://www.elastic.co/guide/en/beats/filebeat/current/filtering-and-enhancing-data.html)
+          for details.
+      - name: tags
+        type: text
+        title: "Tags"
+        description: |-
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/manifest.yml]:
```suggestion
        description: >-
```

File: packages/reconfirm/data_stream/vulnerability/manifest.yml [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+          - forwarded
+          - reconfirm-vulnerability
+      - name: enable_request_tracer
+        type: bool
+        title: "Enable request tracing"
+        default: false
+        multi: false
+        required: false
+        show_user: false
+        description: |-
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/manifest.yml]:
```suggestion
        description: >-
```

File: packages/reconfirm/manifest.yml [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+  - name: reconfirm
+    title: ReConfirm scan results
+    description: Collect attack-surface vulnerability scan results from ReConfirm.
+    inputs:
+      - type: http_endpoint
+        title: Collect ReConfirm scan results via webhook
+        description: Receive scan result webhooks pushed by ReConfirm on scan completion.
+owner:
+  github: elastic/security-service-integrations
+  type: elastic
```
Comment (efd6) [inline packages/reconfirm/manifest.yml]: I don't think this is `type: elastic`. I don't see an enhancement request or issue that is requesting it, so I'd say that it's either a "partner" (would also likely have an issue/planning), or more likely "community" type.

File: packages/reconfirm/data_stream/vulnerability/sample_event.json [hunk 22]
```diff
@@ -0,0 +1,534 @@
...
+                        ],
+                        "tls_version": [
+                            "TLSv1.3"
+                        ]
+                    }
+                ],
+                "page_title": [
+                    "Domain for sale"
+                ],
+                "ssl.expiration": [
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/sample_event.json:277]: This is suspicious. Why do we have dotted keys here?

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml (hunk 11 above)
Comment (m11tch) [reply packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml:19]: This is intentional, it's the standard event.original-preservation pattern (JSE00001) applied across pipelines regardless of input type, specifically for Logstash compatibility: if a document is forwarded via Logstash instead of arriving directly through the Agent's http_endpoint input, it can carry a `message` field that this pipeline is also expected to handle. Other http_endpoint-input packages carry the identical pair for the identical reason — e.g. `bitdefender` (push_configuration), `zoom` (meeting_activity), `cloudflare_logpush` (access_request). Happy to remove it if there's a reason those shouldn't apply here that I'm missing.

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml (hunk 13 above)
Comment (m11tch) [reply packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml:834]: I tried to reproduce this and couldn't. Installed a real
logs-reconfirm.vulnerability@custom pipeline with a marker processor,
then ran a document through the actual pipeline via
_ingest/pipeline/_simulate — the marker came through correctly, so the
hook is invoking the custom pipeline as expected. The syntax also
matches other working packages exactly (e.g. packages/doppler). Could
you share the specific test/repro steps that showed it not working? Happy
to dig further with more to go on.

File: packages/reconfirm/data_stream/vulnerability/sample_event.json (hunk 22 above)
Comment (m11tch) [reply packages/reconfirm/data_stream/vulnerability/sample_event.json:277]: reconfirm.similar_domains.ssl.* is declared in
fields.yml with dotted leaf names (ssl.expiration, ssl.issuer_cn, ...)
under a `nested`-typed parent, and the pipeline sets item.ssl = v.ssl
as a genuine nested object. When retrieved back from Elasticsearch,
the leaf names come back as literal dotted keys rather than a nested
"ssl": {...} object — I confirmed this by generating sample_event.json
from a real indexed document via elastic-package test system
--generate, not hand-writing it. Same thing happens with the
score.base/epss.score/ssl_info.* fields added later in the PR, for
the same reason. Let me know if you'd rather these be typed as fully
nested sub-objects instead of dotted leaf names — happy to restructure
if so, but wanted to flag that this is deliberate/expected as-is, not
an accident.

File: packages/reconfirm/data_stream/vulnerability/_dev/test/pipeline/test-common-config.yml [hunk 23]
```diff
@@ -0,0 +1,5 @@
+fields:
+  tags:
+    - preserve_original_event
```
Comment (efd6) [inline packages/reconfirm/data_stream/vulnerability/_dev/test/pipeline/test-common-config.yml:3]: This is not doing anything since we don't have one under normal circumstances.

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml (hunk 11 above)
Comment (efd6) [reply packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml:19]: Thanks.

This relates to my comment about "preserve_original_event". In a normal (non-logstash) install, a user will be surprised that when they set this option, they do not see an original event in the final document since there is none to start with. There are two options to address this: document when it works, or use `Json.dump` of the complete context into `event.original` if the tag is present.

File: packages/reconfirm/data_stream/vulnerability/sample_event.json (hunk 22 above)
Comment (efd6) [reply packages/reconfirm/data_stream/vulnerability/sample_event.json:277]: Can we use `dot_expander` for these? This would improve the UX.

File: packages/reconfirm/manifest.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (efd6) [reply packages/reconfirm/manifest.yml]: This was reverted. Is there a reason for that? The reversion commit doesn't explain why the decision was made.

File: packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml (hunk 13 above)
Comment (efd6) [reply packages/reconfirm/data_stream/vulnerability/elasticsearch/ingest_pipeline/default.yml:834]: At the time of review, I could see that zero sized objects were present in the documents in the pipeline test expectations. For example, https://github.com/elastic/integrations/pull/20515/changes/BASE..fa32af5c5176c24a5512b7326dc9dfc255062d41#diff-5fa5d243df27642036421f333b152c854ddfbd66782dec5bf5fde5006a9ef957. This appears to be fixed.
